### DISCUSSION TITLE
Custom asdf registry

### DISCUSSION THREAD

**Discussion by Sytten** [jdx/mise#2435] [General]
Hi!
I want to scale usage of `mise` in our organization, but one thing that is kinda problematic is the fact that `asdf` plugins cannot be audited first. Ideally I would like to create my own registry of approved plugins and they should be the only ones that can be installed. Maybe this could be implemented as a backend? (I am not too sure about the concepts).
Open to pitching in the implementation, but I feel this would be a good thing for organizations that want to manage their supply chain. Kinda similar to what [Aqua](https://aquaproj.github.io/) calls a registry.
Thanks

**Comment by mrjk** [comment]
I am looking for the exact same feature. Is it possible to have that overridable somewhere ?

**Comment by jdx** [comment]
https://mise.jdx.dev/configuration.html#mise-shorthands-file-config-mise-shorthands-toml

**Reply by mrjk** [reply]
Ok, thank you very much, I completely missed that one! :+1:

**Reply by NiklasRosenstein** [reply]
I think I'm trying to do the same as @mrjk, but I don't fully understand how the config docs help in that.

Basically, I'd like that I can have my own, separate GitHub repository that acts similar to let's say an aqua registry (i.e. has a `registry.yaml` file) and that I can install tools from it in addition to tools from the official mise registry(/ies).

The [Plugins](https://mise.jdx.dev/plugins.html) docs mention:


> Historically it was the only way to add new tools (as the only backend was [asdf](https://mise.jdx.dev/dev-tools/backends/asdf.html)).
>
> [...] However, now with [core tools](https://mise.jdx.dev/core-tools.html) and backends like [aqua](https://mise.jdx.dev/dev-tools/backends/aqua.html)/[ubi](https://mise.jdx.dev/dev-tools/backends/ubi.html), plugins are no longer necessary to run most tools in mise
>
> If you want to integrate a new tool into mise, you should either try to get it into the [aqua registry](https://mise.jdx.dev/dev-tools/backends/aqua.html) or see if it can be installed with [ubi](https://mise.jdx.dev/dev-tools/backends/ubi.html). Then add it to the [registry](https://github.com/jdx/mise/blob/main/registry.toml).

I only see two options if I don't want to go through contributing the tool to the mise registry:

1. Use a plugin instead (although I'm not sure how I would create a plugin that does not provide just a single tool, but a host of tools through an auqa `registry.yaml`)
2. Fork the mise registry to add my tool and set the `aqua.registry_url` option

Am I getting this right?

Although maybe I'm confusing some things because the option default is `https://github.com/aquaproj/aqua-registry` but there's also the mise [registry.toml](https://github.com/jdx/mise/blob/main/registry.toml) ... 😕

**Comment by iilyak** [comment]
The registry term is a bit overloaded. It could mean aqua registry or mise registry and possibly other things. I am very confused since it seems that original question was answered and people are able to use custom registry. Maybe I am trying to do something different? I want to specify this:

cat registry.yaml
```
❯ cat registry.yaml
packages:
  - name: lua/lua
    type: http
    repo_owner: Lua.org
    repo_name: lua
    url: https://www.lua.org/ftp/lua-{{trimV .Version}}.tar.gz
    description: Lua is a lightweight, high-level, multi-paradigm programming language
    format: tar.gz
```

Also my goal is to configure the registry for project so I don't use `~/.config/mise/config.toml` file. Here is how I try to do it

cat 
```
❯ cat mise.toml
[tools]
"aqua:lua" = 'latest'

[settings.aqua]
"registry_url" = "registry.yaml"
```

I got the following error

```
mise WARN  invalid aqua tool: lua
mise cloning aqua registry to "/Users/iilyak/Library/Caches/mise/aqua-registry"
mise WARN  git clone failed: unsupported URL protocol; class=Net (12)
error: '/Users/iilyak/Code/experiments/mise-test/registry.yaml' does not look like a v2 or v3 bundle file
fatal: invalid gitfile format: /Users/iilyak/Code/experiments/mise-test/registry.yaml
fatal: Could not read from remote repository.
```

I read it as `registry_url` has to point to real repository. How would I use a plane `registry.yaml` file? Is what I ask different from the discussion title (if so should I start a new discussion)?

**Comment by JamBalaya56562** [comment]
@iilyak @NiklasRosenstein — this is answerable now, and neither of the workarounds you two landed on is needed.

The setting is **`aqua.registries`**. `aqua.registry_url`, which you both found, is the older single-source form and is now deprecated (removed in 2027.12.0). The difference matters here: `registry_url` takes a **repository**, while `aqua.registries` accepts a repository URL, **a direct URL to a `registry.yaml`/`registry.yml`**, or an absolute **`file://`** URL to a local directory or file.

@iilyak, that is the direct answer to *"How would I use a plain `registry.yaml` file?"* — your `registry_url = "registry.yaml"` cannot work because a bare relative filename is not a URL. On **v2026.8.3** it now fails with a clearer message than the git-clone one you got:

```console
mise WARN  ... invalid aqua registry URL registry.yaml: relative URL without a base
```

With an absolute `file://` and your registry unchanged:

```toml
[settings]
aqua.registries = ["file:///Users/iilyak/Code/experiments/mise-test/registry.yaml"]
```

```console
$ mise ls-remote aqua:myorg/mytool
2.95.0
2.96.0
2.97.0
```

There is a second, separate problem in your snippet: `[tools] "aqua:lua"`. aqua package names are `owner/repo`, and your registry defines `name: lua/lua`, so it has to be `aqua:lua/lua`. That is what produced your `mise WARN invalid aqua tool: lua` — I reproduced the identical warning with a single-segment name. And settings work per-project, so keeping this in the project's `mise.toml` rather than `~/.config/mise/config.toml` is fine, which was your other requirement.

@NiklasRosenstein — your option 2 (fork the mise registry) is not necessary. Point `aqua.registries` at **your own repository** containing a `registry.yaml`; it is consulted *before* the baked-in registry, and with `aqua.baked_registry` left enabled the official one stays as a fallback, which is the "in addition to" you asked for. On the confusion at the end of your comment: mise's [`registry.toml`](https://github.com/jdx/mise/blob/main/registry.toml) maps short names like `mage` to a backend; aqua's `registry.yaml` describes how to fetch a package. `aqua.registries` concerns the second one only.

@Sytten — for the original ask, "my own registry of approved plugins and **only** those installable", add `aqua.baked_registry = false`. Measured: with it off, a tool that exists in the baked registry stops resolving:

```console
$ mise ls-remote aqua:magefile/mage
mise WARN  ... no aqua-registry found for magefile/mage
```

so only what your registry declares can be installed through aqua. `disable_backends` complements this if you also want to stop `asdf:` being usable at all.

One aside, since this thread began with asdf plugins: `shorthands_file`, which is what the docs link above describes, is itself deprecated now (warns from 2026.6.0, removed 2026.12.0) in favour of `[plugins]` in `mise.toml` — same `name = "url"` rows.

**Reply by risu729** [reply]
The aqua registries are different from mise registry itself. Users should use `tool_aliases` if they want a custom registry. mise doens't offer configs to extend/replace the mise registry.
